### ISSUE TITLE
Problem with reproducing results

### ISSUE THREAD

**Issue by kacperkan** [A2Zadeh/Variational-Autodecoder#1]
Hi

I struggle with achieving compelling qualitative results while reconstructing data on MNIST dataset.

Run commands:
1. Training the model:
```bash
 python pytorch/train_model.py train mnist generated 0 --cuda --kl --model vae_free
```
2. Plotting the results:
```bash
python scripts/plot_mnist.py mnist generated 0
```

Since you didn't share the code that preprocesses the data, I implemented it on my own (the code is available in my fork of this repo, under the file `pytorch/make_datasets.py`, <[link](https://github.com/kacperkan/Variational-Autodecoder/blob/master/pytorch/make_datasets.py)>

I tried both:
- normalizing MNIST to the range [0, 1] (hence, adapting `plot_mnist.py` accordingly)
- leaving MNIST in the range [0, 255]

Below, I share the config produced with the `make_configs.py` scripts.
```json
{
    "n_latent_params": 25,
    "n_hidden": 200,
    "n_hidden_layers": 2,
    "use_adam": true,
    "dropout": null,
    "layer_norm": false,
    "latent_param_lr": 0.001,
    "net_lr": 0.001,
    "test_latent_param_lr": 0.001,
    "masked": 0.1,
    "latent_param_init": "pca",
    "test_latent_param_init": "normal",
    "activation_fn": "relu",
    "config_folder": "generated",
    "config_num": 0
}
```
I met the same problem with the config `10.json` produced by that script.
Some obtained reconstructions (values of the MNIST were normalized to [0, 255]):

Input | Reconstruction
----|----
![in0](https://user-images.githubusercontent.com/57393775/69537235-fdb54500-0f7f-11ea-94dc-5cbe09eef336.jpg) |  ![out0](https://user-images.githubusercontent.com/57393775/69537255-0d348e00-0f80-11ea-9f0c-664e88289aad.jpg)
![in1](https://user-images.githubusercontent.com/57393775/69537300-24737b80-0f80-11ea-85ba-223a252eb87e.jpg) |  ![out1](https://user-images.githubusercontent.com/57393775/69537311-2a695c80-0f80-11ea-9803-99d0ef8e0549.jpg)

In general, all reconstructions look the same.

**Question:** Is there a problem in the original code or is my preprocessing wrong? If the first thing, then do you know where the problem is located in the code?

**Comment by kacperkan** [comment]
I solved the problem by fixing a vector that is sampled during the forward inference (call of `forward`), so the same vector is used during the reparametrization trick. So the implementation of the `MaskedVAD` would look like this:
```python
class MaskedVAD(MaskedDecoder):
    def __init__(self, *args, log_var=0.0):
        super(MaskedVAD, self).__init__(*args)

        print("log var: {}".format(log_var))
        log_var_param = nn.Parameter(
            data=torch.tensor(log_var), requires_grad=False
        )
        self.register_parameter("log_var", log_var_param)
        self.std = np.exp(log_var / 2)
        self.eps = nn.Parameter(torch.Tensor(1, self.latent_size), requires_grad=False)  # fixing latent code, since the latent means / stds are learnable
        nn.init.normal_(self.eps) # sampling from N(0, 1)
        self.verbose = False

    def set_verbose(self, verb):
        self.verbose = verb

    def forward(self, latents):
        x = latents + self.eps * self.std  # reparametrization
        for layer in self.hiddens:
            x = layer(x)

        return x
```
However, I'm not sure then whether it is mathematically correct. Would you mind describing it is a correct solution? If not, how should the convergence problem be fixed?

**Comment by Dai-ge** [comment]
Hi kacperkan,
Did you solve this problem now? It seems like using "a fixed latent code" method you show here will make the variational auto-decoder method move back to the naive auto-decoder. Have you tried other hyper-parameter settings?
